src/main/java/org/spongepowered/api/item/recipe/crafting/ShapedCraftingRecipe.java

```diff
@@ -0,0 +1,229 @@
+                                        " one for each symbol in the aisle!"));
+
+                        if(!ingredientPredicate.test(itemStack.createSnapshot()))
+                            continue byShiftingTheAisle;
+                    }
+                }
+
+                // Make sure the gap is empty
+                for(int gapX = 0; gapX < gapWidth; gapX++) {
+                    for(int gapY = 0; gapY < gapHeight; gapY++) {
+                        int gridX = gapX + (gapX >= offsetX ? getWidth() : 0);
+                        int gridY = gapY + (gapY >= offsetY ? getHeight() : 0);
+                        boolean empty = grid.getSlot(gridX, gridY)
+                                .flatMap(Slot::peek)
+                                .map(itemStack -> itemStack.getItem() == ItemTypes.NONE)
+                                .orElse(true);
+
+                        if(!empty)
+                            continue byShiftingTheAisle;
+                    }
+                }
+
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    /**
+     * Gets the width of the grid this ShapedCraftingRecipe fits into.
+     *
+     * @return The width of the grid
+     */
+    int getWidth();
+
+    /**
+     * Gets the height of the grid this ShapedCraftingRecipe fits into.
+     *
+     * @return The height of the grid
+     */
+    int getHeight();
+
+    @Override
+    default int getRecipeSize() {
+        return getWidth() * getHeight();
+    }
+
+    interface Builder extends ResettableBuilder<ShapedCraftingRecipe, Builder> {
+        /**
+         * Sets the aisle pattern for the shaped recipe.
+         *
+         * @param aisle A string array of ingredients
+         * @return The builder
+         */
+        Builder aisle(String... aisle);
+
+        /**
+         * Sets an ingredient predicate based on the aisle pattern for the shaped recipe.
+         *
+         * @param symbol The ingredient symbol
+         * @param ingredient The ingredient predicate to set, or remove if null
+         * @return The builder
+         * @throws IllegalArgumentException If the aisle does not contain
+         *     the specified character symbol
+         */
+        Builder where(char symbol, @Nullable Predicate<ItemStackSnapshot> ingredient) throws IllegalArgumentException;
+
+        /**
+         * Sets an ingredient predicate based on the aisle pattern for the shaped recipe.
+         * This mimics the vanilla checking behavior.
+         *
+         * @param symbol The ingredient symbol
+         * @param ingredient The ingredient to set, or remove if null
+         * @return The builder
+         * @throws IllegalArgumentException If the aisle does not contain
+         *     the specified character symbol
+         */
+        Builder where(char symbol, @Nullable ItemStackSnapshot ingredient) throws IllegalArgumentException;
```

Comment:
@Limeth: Thank link seems to be broken. Can the implementation take care of passing in `ItemStackSnapshot.NONE` to the predicate, if that's what the issue is?